Implement hook module running external commands

clems4ever opened the pull request:
Instead of handling the hook logic in the Mesos process, we spawn a new
process running the command defined in parameters, passes the inputs as JSON string and get the
output as JSON string. That way, the task is isolated from the mesos
agent.

clems4ever committed:
Implement hook module running external commands

Instead of handling the hook logic in the Mesos process, we spawn a new
process running the command, passes the inputs as JSON string and get the
output as JSON string. That way, the task is isolated from the mesos
agent.

pierresouchay commented on src/CommandHook.cpp:
```diff
@@ -0,0 +1,89 @@
+  const ::mesos::SlaveInfo& slaveInfo) {
+  if(m_runTaskLabelCommand.empty()) return None();
+
+  LOG(INFO) << "Calling external command for slaveRunTaskLabelDecorator";
```
I would also display the command we are launching

clems4ever replied:
Done

pierresouchay commented on src/CommandHook.cpp:
```diff
@@ -0,0 +1,89 @@
+  const ::mesos::ExecutorInfo& executorInfo) {
+  if(m_executorEnvironmentCommand.empty()) return None();
+
+  LOG(INFO) << "Calling external command for slaveExecutorEnvironmentDecorator";
```
same thing, display which command

clems4ever replied:
Done

pierresouchay commented on src/CommandRunner.cpp:
```diff
@@ -0,0 +1,114 @@
+  TemporaryFile() {
+    char filepath[] = TEMP_FILE_TEMPLATE;
+    if(mkstemp(filepath) == -1)
+      throw std::runtime_error("Unable to create temporary file");
```
In the message, explain that this error is related to the module: ie: "Unable to create a temporary file to run commands" => so the administrators knows that error might be related to this part of code, not in any other part of Mesos

clems4ever replied:
Well, in the logs you will have the filename of the source code anyway but I will still augment the log line because it might not be easy to know if this file is in Mesos or somewhere else.

pierresouchay commented on src/CommandRunner.cpp:
```diff
@@ -0,0 +1,114 @@
+    std::ofstream ofs;
+    ofs.open(m_filepath);
+    ofs << content;
+    std::flush(ofs);
```
Forgot to close the file? Are your sure the destructor to close the fd ?

clems4ever replied:
Yes, it is automatically closed: https://stackoverflow.com/questions/12900016/does-ofstream-close-its-files-automatically. I will call it though so that it's not even questionable anymore

pierresouchay commented on src/CommandRunner.cpp:
```diff
@@ -0,0 +1,114 @@
+
+  std::array<char, 128> buffer;
+  std::string stdoutData;
+  std::shared_ptr<FILE> pipe(popen(fullCommand.str().c_str(), "r"), pclose);
```
I suspect all of this being part of C++11... I assume pclose is called on destruction

clems4ever replied:
Yes, it wraps the ptr into a smart pointer and call pclose when the pointer is not referenced anymore, i.e., out of scope in this case.

pierresouchay commented on src/CommandRunner.cpp:
```diff
@@ -0,0 +1,114 @@
+  if (!pipe)
+    throw std::runtime_error("popen() failed!");
+
+  while (!feof(pipe.get())) {
```
well... this must work, but I'll add a timeout for the execution of all of this.
For instance, I am curious of what is doing a simple script such as `cat <&0`... I suspect it never ends

While reading, we should have a timeout, and kill the task when timeout happens

clems4ever replied:
Done.

pierresouchay reviewed (commented):
That's a very good start, but I think we should enforce a bit the babysitting of process and handling properly timeouts

clems4ever committed:
Add more logs and explicitely close file streams

clems4ever committed:
Add a timeout preventing mesos-agent to be blocked by long tasks

clems4ever committed:
Remove useless ruby dependency

clems4ever committed:
Add .travis.yml to build and test the modules

fboismenu commented on examples/docker/modules/slaveExecutorEnvironmentDecorator.sh:
```diff
@@ -0,0 +1,17 @@
+
+cat $INPUT_FILE
+
+read -r -d '' OUTPUT << EOM
```
why not a more direct cat << EOF > $OUTPUT_FILE
...
EOF
?

clems4ever replied:
Will do

kamaradclimber commented on src/CommandHook.cpp:
```diff
@@ -0,0 +1,92 @@
+
+  auto outputJsonTry = JSON::parse(output);
+  if(outputJsonTry.isError()) {
+    LOG(WARNING) << "Error when parsing string to JSON.";
```
OPT: isn't it possible to show the error?

clems4ever replied:
Will do.

pierresouchay replied:
+1 show the payload

fboismenu commented on src/CommandHook.hpp:
```diff
@@ -0,0 +1,96 @@
+   * @param removeExecutorCommand The command used by
+   * slaveRemoveExecutorHook if provided.
+   */
+  CommandHook(const std::string& runTaskLabelCommand,
```
Could have used a map<string,string> here ... easier to inspect, validate, etc

clems4ever replied:
FAR

Mwea commented on src/CommandHook.cpp:
```diff
@@ -0,0 +1,92 @@
+    return Error("Malformed JSON");
+  }
+
+  auto outputJson = outputJsonTry.get();
```
Would

``` 
auto outputJson = outputJsonTry.get();

if(!outputJson.is<JSON::Object>()) {
  return Error("Malformed Protobuf");  
}

auto proto = ::protobuf::parse<Proto>(outputJson);
if(proto.isError())
  LOG(WARNING) << "Error when parsing JSON to protobuf";
return proto;

``` 
 be better if you want to avoid unnecessary None() return ?

clems4ever replied:
Will do

kamaradclimber commented on src/CommandRunner.cpp:
```diff
@@ -0,0 +1,204 @@
+#include <glog/logging.h>
+#include <stout/try.hpp>
+
+#define TEMP_FILE_TEMPLATE "/tmp/criteo-mesos-XXXXXX"
```
FAR: I think we should use a better location that bare /tmp.

Are you sure that this directory is cleaned automatically (in case your program fail to do so)?

Configuration on centos7 seems to be 
```
v /tmp 1777 root root 10d
v /var/tmp 1777 root root 30d
```
(for systemd builtin cleaning tmp files)

clems4ever replied:
What does "better" mean?

fboismenu commented on src/CommandRunner.cpp:
```diff
@@ -0,0 +1,204 @@
+
+#define TEMP_FILE_TEMPLATE "/tmp/criteo-mesos-XXXXXX"
+
+#define READ 0
```
very opt: enum?

clems4ever replied:
Won't do or FAR.

clems4ever replied:
But you are right.

kamaradclimber commented on src/CommandRunner.cpp:
```diff
@@ -0,0 +1,204 @@
+    close(p_stdout[READ]);
+    dup2(p_stdout[WRITE], WRITE);
+    execl("/bin/sh", "sh", "-c", command, NULL);
+    LOG(ERROR) << "Error when executing command \"" << command << "\"";
```
isn't possible to have the error?
newbie question: do we know the type of condition where this code would be reached?

clems4ever replied:
It is possible to have the exit code of the command. Regarding the reachability of LOG(ERROR), it can only be executed if execl fails as stated in the doc https://linux.die.net/man/3/execl, ```The exec() functions only return if an error has occurred```.
I will display the exit code.

pierresouchay replied:
Replace execl("/bin/sh", "sh", "-c", command, NULL); by 

execl(path_to_script, path_to_script, input_file, output_file)

kamaradclimber commented on src/CommandRunner.cpp:
```diff
@@ -0,0 +1,204 @@
+
+  // executed in parent
+  if (infp == NULL)
+    close(p_stdin[WRITE]);
```
OPT: style comment, I think adding brackets would ease future refactoring and prevent future mistakes

clems4ever replied:
Will do

fboismenu commented on src/CommandRunner.cpp:
```diff
@@ -0,0 +1,204 @@
+    if (rc < 0) {
+      LOG(ERROR) << "Error when waiting for child process running the "
+                 << "external command";
+      return;
```
You may want to kill it here as well

clems4ever replied:
Normally if we have rc == -1, this is because there is nothing to wait for, then nothing to kill. But I will still try to kill anyway to be sure.

kamaradclimber commented on src/CommandRunner.cpp:
```diff
@@ -0,0 +1,204 @@
+void runCommandWithTimeout(const std::string& command, int timeout) {
+  int status;
+  int times = 0;
+  const int SECONDS = 1000000000;
```
seems very close to 2**31, I would recommend to be safer here

clems4ever replied:
Indeed! Fixing it.

kamaradclimber commented on src/CommandRunner.cpp:
```diff
@@ -0,0 +1,204 @@
+
+  if(times == max_times) {
+    LOG(WARNING) << "External command took too long to exit. Killing it...";
+    kill(pid, SIGKILL);
```
how do we check that kill was successful?

clems4ever replied:
Will check

pierresouchay replied:
We should do first a SIGINT... wait a bit and send a SIGKILL then

clems4ever replied:
Done

fboismenu commented on tests/CommandHookTest.cpp:
```diff
@@ -0,0 +1,168 @@
+using namespace criteo::mesos;
+
+std::string bash(const std::string& script) {
+  return "/bin/bash " + resources_path + script;
```
Command Executor is using sh ?

clems4ever replied:
Changed with @pierresouchay's comment.

Mwea commented on src/CommandRunner.cpp:
```diff
@@ -0,0 +1,204 @@
+   * @return The content of the file.
+   */
+  std::string readAll() const {
+    std::ifstream ifs(m_filepath);
```
Even if the file was accessible during the instantiation of TemporaryFile, the file could have been deleted or even chmod... May need to check for errors here ?

clems4ever replied:
Yes, I had this in mind as well. I will implement it in another review.

Mwea commented on src/CommandRunner.cpp:
```diff
@@ -0,0 +1,204 @@
+   */
+  void write(const std::string& content) const {
+    std::ofstream ofs;
+    ofs.open(m_filepath);
```
should also check for failure here ?

clems4ever replied:
see comment above

Mwea commented on src/CommandRunner.cpp:
```diff
@@ -0,0 +1,204 @@
+ * finish before the timeout deadline.
+ *
+ * @param command The command to execute in the child process.
+ * @param timeout The timeout deadline in seconds before killing the
```
Define the unit of timeout ?

clems4ever replied:
What do you mean? it's declared in seconds here.

Mwea commented on src/CommandRunner.cpp:
```diff
@@ -0,0 +1,204 @@
+  int p_stdin[2], p_stdout[2];
+  pid_t pid;
+  if (pipe(p_stdin) != 0 || pipe(p_stdout) != 0)
+    return -1;
```
Add brackets to prevent futur refactor to fails

clems4ever replied:
Will do

pierresouchay commented on src/CommandHook.cpp:
```diff
@@ -0,0 +1,92 @@
+  if(outputJson.is<JSON::Object>()) {
+    auto proto = ::protobuf::parse<Proto>(outputJson);
+    if(proto.isError())
+     LOG(WARNING) << "Error when parsing JSON to protobuf";
```
Error when parsing JSON to protobuf => error while converting JSON to protobuf

clems4ever replied:
Will do

clems4ever committed:
Improve logging and do not fork a bash to run external command

Also add some parentheses around if blocks and try sending a SIGTERM before
SIGKILL to give the child the possibility to clean up peacefully.

clems4ever committed:
Update the README with design choices

clems4ever commented:
I'm done with the changes.
